### ISSUE TITLE
MethodHandles are not thread-safe: invoking different MethodHandles concurrently leads to ClassCastException

### ISSUE THREAD

**Issue by yrodiere** [oracle/graal#3346]
**Describe the issue**

Invoking `MethodHandles` concurrently apparently leads to an internal mix-up; ultimately the "self" argument passed to `MethodHandle.invoke` will be cast to the type expected by *a different `MethodHandle` being invoked in another thread*.

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. `git clone https://github.com/yrodiere/graalvm-methodhandle.git`
2. `git switch concurrency`
3. `./reproducer.sh`

The script will build a JAR, then execute the reproducer in a JVM (it will work), then generate the native image (`native-image` must be in the path), then execute the reproducer using the native image. You will see several stack traces similar to the one I gave below in "More details".

**Describe GraalVM and your environment:**
 - GraalVM version: latest snapshot (https://github.com/oracle/graal/commit/bec1cfb60d9d6988fd2307b2337b5d53fcf567f7) or 21.0.0.2
 - JDK major version: 11
 - OS: Fedora 32, `Linux yrodiere.redhat 5.11.12-100.fc32.x86_64 #1 SMP Wed Apr 7 16:52:03 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
 - Architecture: AMD64

**More details**

For example, let's say I have two `MethodHandles`, `MH1` and `MH2`, where `MH1` expects to be called on objects of type `Target1` and `MH2` expects to be called on objects of type `Target2`. Running these two `MethodHandles` repeatedly and concurrently, at some point I will get an exception like this:

```
java.lang.IllegalStateException: Exception while invoking 'MethodHandle(Target1)Object' on 'org.hibernate.testing.graalvm.methodhandle.Target1@4fc25737': org.hibernate.testing.graalvm.methodhandle.Target2 cannot be cast to org.hibernate.testing.graalvm.methodhandle.Target1
        at org.hibernate.testing.graalvm.methodhandle.Main.invokeManyTimes(Main.java:69)
        at org.hibernate.testing.graalvm.methodhandle.Main.lambda$main$0(Main.java:43)
        at java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(Thread.java:834)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: java.lang.ClassCastException: org.hibernate.testing.graalvm.methodhandle.Target2 cannot be cast to org.hibernate.testing.graalvm.methodhandle.Target1
        at com.oracle.svm.methodhandles.MethodHandleIntrinsic.execute(MethodHandleIntrinsic.java:285)
        at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:80)
        at com.oracle.svm.methodhandles.MethodHandleIntrinsic.execute(MethodHandleIntrinsic.java:175)
        at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:80)
        at java.lang.invoke.LambdaForm$NamedFunction.invokeWithArguments(LambdaForm.java:74)
        at java.lang.invoke.LambdaForm.interpretName(LambdaForm.java:981)
        at java.lang.invoke.LambdaForm.interpretWithArguments(LambdaForm.java:958)
        at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:171)
        at com.oracle.svm.methodhandles.MethodHandleIntrinsic.execute(MethodHandleIntrinsic.java:175)
        at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:80)
        at java.lang.invoke.LambdaForm$NamedFunction.invokeWithArguments(LambdaForm.java:74)
        at java.lang.invoke.LambdaForm.interpretName(LambdaForm.java:981)
        at java.lang.invoke.LambdaForm.interpretWithArguments(LambdaForm.java:958)
        at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:171)
        at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:0)
        at java.lang.invoke.LambdaForm$MH/658475348.invoke_MT(LambdaForm$MH)
        at org.hibernate.testing.graalvm.methodhandle.Main.invokeManyTimes(Main.java:66)
        ... 7 more
```

As you can see, the failing cast from `Target2` to `Target1` does not make sense, since the MethodHandle expects a `Target1` and the object is of type `Target1`. So there must be some mixup internally.

The problem may be caused by `com.oracle.svm.methodhandles.MethodHandleIntrinsic#cache`, which is a static cache with no protection against concurrent access whatsoever, but that's just a wild guess.

**Comment by loicottet** [comment]
The failure seems to be corrected on master and on the upcoming 21.1 release. However the fact that the cache is not thread-safe is something that should be corrected, thanks for the catch!

**Comment by loicottet** [comment]
Fixed by 45c2408c5b6a570c5acac199c147bdf08acf74c1
